core/command_line_parser.cpp

```diff
@@ -619,10 +618,6 @@ Error CommandLineParser::parse(const PoolStringArray &p_args) {
 
 	_read_default_args();
 
-	return OK;
-}
-
-Error CommandLineParser::validate() {
 	for (int i = 0; i < _options.size(); ++i) {
```

Comment:
You will not able to define a help method and required arguments. This is how it done in other parsers.

Reply:
I haven't removed validation code, but merged it into `parse()`. As a user, I still don't quite get the usefulness of this to be honest. Could you elaborate?

Reply:
Yes I understand. But without this separation, if you declare a `help` command and mark some other command as required, the help command will not work because it will fail on required check.

Reply:
With this separation you can check for the help / version commas first and then call `validate` to check if other arguments are valid.

Reply:
> Yes I understand. But without this separation, if you declare a `help` command and mark some other command as required, the help command will not work because it will fail on required check.

Alright, yeah I understand this now. But is it only useful for `help`/`version` options, or are there other use cases where this separation may be required?

Reply:
I know only about this one. This is how it implemented in `Boost.program_options`.

Reply:
Yeah, I'm trying to be a little bit pragmatic here, what works for other libraries in particular environments may fail short under other use cases, and there are other features that are not currently implemented, like allowing unrecognized options.

If this is only/mostly useful for those `help`/`version` options, perhaps it may be just enough to allow to skip validation checks if certain options are detected on the command-line.

Reply:
> and there are other features that are not currently implemented, like allowing unrecognized options.

For me, the `validate` approach is much more flexible. Shouldn't be a problem.

> If this is only/mostly useful for those help/version options, perhaps it may be just enough to allow to skip validation checks if certain options are detected on the command-line.

For me personally, another switch in option to ignore validation step if specified will be less obvious. But it's up to you.

Reply:
> For me, the `validate` approach is much more flexible. Shouldn't be a problem.

As they say, flexibility comes with a cost of added maintenance burden. 😛

> For me personally, another switch in option to ignore validation step if specified will be less obvious. But it's up to you.

That's not what I propose actually. I'm saying that `parse()` itself could detect those `help`/`version` options. If they are set, validation step will be simply skipped.

I'm also torn by the signal which is emitted upon parsing/validation. If you look at the current code:
```gdscript
	for (int i = 0; i < _options.size(); ++i) {
		CommandLineOption *option = _options.get(i).ptr();
		const Map<const CommandLineOption *, PoolStringArray>::Element *E = _parsed_values.find(option);
		if (E) {
			option->emit_signal("parsed", E->value());
		}
	}
```

you'll see that there's consistency: what we find under `_parsed_values` get parsed, so we emit `parsed` signal. IIRC, in the original implementation `parsed` signal was emitted after `validate()/notify()`. The problem is that whether we rename the signal or the method, it's going to be inconsistent. That's another reason why I've initially decided to merge parse and validation checks into a single `parse()` method.

Of course, with current implementation, there's still an issue that `--help` will not print anything if required options are found, as you said, so this needs to be fixed regardless. I'm trying to find a solution which does not suffer from inconsistencies/limitations, and one of the major inconsistency with the API that  has both `parse()` and `validate()` is signals.

I know I proposed to rename `parsed` signal to `validated` in your original PR, but after digging through implementation now, I got concerned with it again. If we bring back `validate()` method, the signal should also be renamed to `validated` in `CommandLineOption`.

This suggests to me, perhaps, `CommandLineOption` could have both `parsed` and `validated` signals if we go this route.

Reply:
> That's not what I propose actually. I'm saying that parse() itself could detect those help/version options. If they are set, validation step will be simply skipped.

But how? I don't really like the idea of ​​having a hardcoded check for `version` and `help` text inside.

But understand your opinion about signals, yes.

Reply:
> But how? I don't really like the idea of ​​having a hardcoded check for `version` and `help` text inside.

Like this (only one line of condition is added):
```c++
if (find_option("help").is_null() && find_option("version").is_null()) {
	for (int i = 0; i < _options.size(); ++i) {
		const CommandLineOption *option = _options[i].ptr();
		if (unlikely(option->is_required() && !_parsed_values.has(option))) {
			_error_text = vformat(RTR("Option '%s' is required but missing."), _to_string(option->get_names()));
			return ERR_PARSE_ERROR;
		}
	}
}
```

I also think that this could be hack-ish, but that's the best solution to the problem if and only if `help`/`version` are the only use cases that require skipping validation for required options. So far, neither me nor you have come up with other use cases where a more sophisticated approach would be needed.

Reply:
> hardcoded check for `version` and `help` text inside.

It may not be necessary to hardcode text. As I said, we have both `add_help_option()` and `add_version_option()` in public API. Those could be treated as special options and be recognized accordingly by the parser.

Reply:
> Like this:

Then the text will be hardcoded :)

> As I said, we have both add_help_option() and add_version_option() in public API. Those could be treated as special options and be recognized accordingly by the parser.

Then options will have a switch (maybe internal) to skip validation as I mentioned before.

As to me that this behavior (with automatic validation skip) will be less obvious then a direct call to detect required variables. But it just me.

Reply:
I've personally never used `Boost.program_options`, but I did use Python's `argparse`. There, if you define an option which is `required=True`, then somehow it still allows to parse `--help` and it prints a message, so I assume Python does have an internal mechanism to handle such cases. Also, according to [`argparse.required`](https://docs.python.org/3/library/argparse.html#required) documentation:

> Required options are generally considered bad form because users expect options to be optional, and thus they should be avoided when possible.

So having API which allows to explicitly validate frowned upon approach for defining options might not be a good idea.

As I said, `parse()` already does a lot of validation checks for options/arguments prior to `validate()` (see various internal `_validate_*` methods), so it's also not 100% clear which checks are purely user-facing to make a better distinction (those that simply set `_error_text` or just `ERR_FAIL_*`).

I'd personally like to keep API simpler if there's a possibility to make it simpler without losing functionality. If users say that `validate()` is needed, it could be added in future versions without breaking compatibility, I guess.

Reply:
Hmm... This makes.
On second thought a command option to skip validation if the command is specified not that bad.

Reply:
I have added ability to mark options as "meta", see recent commits. Options that are meta are designed to get information about application itself, so if any such option is detected on the command-line, there's no need to check for required options, because meta options are not designed to interact with user options.

Reply:
Awesome!
I would allow user to set it, cost nothing, but it's up to you.